File: cos/olympus/src/main/java/cos/olympus/game/events/Damage.java
package cos.olympus.game.events;

import cos.olympus.game.Creature;

public record Damage(
        int id,
        int tick,
        Creature victim,
        Spell spell,
        int amount,
        boolean crit
) implements Event {

    public cos.ops.out.Damage toUserOp(int userId) {
        return new cos.ops.out.Damage(id, tick, userId, spell.source().id(), victim.id(), amount, spell.id(), crit);
    }

    @Override public String toString() {
        return "Damage{" +
                "id=" + id +
                ", tick=" + tick +
                ", victim=" + victim.id() +
                ", spell=" + spell.id() +
                ", amount=" + amount +
                '}';
    }
}

File: cos/src/main/java/cos/olympus/game/events/Damage.kt
package cos.olympus.game.events

import cos.olympus.game.Actor
import cos.olympus.game.TickId
import cos.ops.out.Damage

data class Damage(
    override val id: Int,
    override val tick: TickId,
    val victim: Actor,
    val spell: Spell,
    val amount: Int,
    val crit: Boolean
) : Event {

    fun toUserOp(userId: Int) = Damage(id, userId, spell.source.id, victim.id, amount, spell.id, crit)
}

